rust/blockstore/src/arrow/sparse_index.rs

```diff
@@ -361,213 +361,99 @@ impl SparseIndexReader {
-                // block with this prefix.
-                if non_start_curr_key.is_some()
-                    && prefix == non_start_curr_key.unwrap().prefix.as_str()
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-                // If prefix is between the current delim and next delim then there could
-                // be keys in this block that have this prefix.
-                if (non_start_curr_key.is_none()
-                    || prefix > non_start_curr_key.unwrap().prefix.as_str())
-                    && (prefix <= non_start_next_key.unwrap().prefix.as_str())
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-            } else {
-                // Last block.
-                if non_start_curr_key.is_none()
-                    || prefix >= non_start_curr_key.unwrap().prefix.as_str()
-                {
-                    block_ids.push(curr_block_value.id);
-                }
-            }
-        }
-        block_ids
-    }
-
-    /// Get all block ids that have keys with the given prefix and key greater than the given key
-    pub(super) fn get_block_ids_gt<'a, K: ArrowReadableKey<'a> + Into<KeyWrapper>>(
+    pub(super) fn get_block_ids_range<'prefix, 'referred_data, K, PrefixRange, KeyRange>(
         &self,
-        prefix: &str,
-        key: K,
-    ) -> Vec<Uuid> {
-        let data = &self.data;
-        let forward = &data.forward;
-        let curr_iter = forward.iter();
-        let mut next_iter = forward.iter().skip(1);
-        let mut block_ids = vec![];
-        for (curr_delim, curr_block_value) in curr_iter {
-            let curr_key = match curr_delim {
-                SparseIndexDelimiter::Start => None,
-                SparseIndexDelimiter::Key(k) => Some(k),
-            };
-            let mut next_key: Option<&CompositeKey> = None;
-            if let Some((next_delim, _)) = next_iter.next() {
-                next_key = match next_delim {
-                    SparseIndexDelimiter::Start => {
-                        panic!("Invariant violation. Sparse index is not valid.")
-                    }
-                    SparseIndexDelimiter::Key(k) => Some(k),
+        // These key ranges are flattened instead of using a single RangeBounds<CompositeKey> because not all keys have a well-defined min and max value. E.x. if the key is a string, there would be no way to get the range for all keys within a specific prefix.
+        prefix_range: PrefixRange,
+        key_range: KeyRange,
+    ) -> Vec<Uuid>
+    where
+        K: ArrowReadableKey<'referred_data>,
+        PrefixRange: RangeBounds<&'prefix str>,
+        KeyRange: RangeBounds<K>,
+    {
+        let forward = &self.data.forward;
+
+        // We do not materialize the last key of each block, so we must check the next block's start key to determine if the current block's end key is within the query range.
+        forward
```

Comment:
this is quite hard to read. In general, I wouldn't prefer deep nested chaining of this fashion. Not sure what the rust guidelines are though. Would prefer easier to understand code at the cost of more verbosity

Reply:
Resolved offline